ci-bench/src/main.rs

```diff
@@ -0,0 +1,267 @@
+
+            // Output results in CSV (note: not using a library here to avoid extra dependencies)
+            for (name, instr_count) in results {
+                println!("{name},{instr_count}");
+            }
+        }
+        Command::RunSingle { index, side } => {
+            bench_lib::run_single(index, side, &benchmarks)?;
+        }
+        Command::Compare {
+            baseline_input,
+            candidate_input,
+        } => {
+            bench_lib::compare(baseline_input, candidate_input)?;
+        }
+    }
+
+    Ok(())
+}
+
+/// Returns all benchmarks wrapped in a [`BenchmarkCollection`]
+fn get_benchmarks() -> anyhow::Result<BenchmarkCollection> {
+    let mut benchmarks = Vec::new();
+    for param in all_benchmark_params() {
+        add_benchmark_group(&mut benchmarks, param);
+    }
+
+    BenchmarkCollection::new(benchmarks)
+}
+
+/// Returns the benchmark params to use for each group of benchmarks
+fn all_benchmark_params() -> Vec<BenchmarkParam> {
+    vec![
+        BenchmarkParam::new(
+            KeyType::Rsa,
+            rustls::cipher_suite::TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
+            &rustls::version::TLS12,
+            "1.2_rsa_aes",
+        ),
+        BenchmarkParam::new(
+            KeyType::Rsa,
+            rustls::cipher_suite::TLS13_AES_128_GCM_SHA256,
+            &rustls::version::TLS13,
+            "1.3_rsa_aes",
+        ),
+        BenchmarkParam::new(
+            KeyType::Ecdsa,
+            rustls::cipher_suite::TLS13_AES_128_GCM_SHA256,
+            &rustls::version::TLS13,
+            "1.3_ecdsa_aes",
+        ),
+        BenchmarkParam::new(
+            KeyType::Rsa,
+            rustls::cipher_suite::TLS13_CHACHA20_POLY1305_SHA256,
+            &rustls::version::TLS13,
+            "1.3_rsa_chacha",
+        ),
+        BenchmarkParam::new(
+            KeyType::Ecdsa,
+            rustls::cipher_suite::TLS13_CHACHA20_POLY1305_SHA256,
+            &rustls::version::TLS13,
+            "1.3_ecdsa_chacha",
+        ),
+    ]
+}
+
+/// Adds a group of benchmarks for the specified parameters
+///
+/// The benchmarks in the group are:
+///
+/// - Handshake without resumption
+/// - Handshake with session id resumption
+/// - Handshake with ticket resumption
+/// - Transfer a 1MB data stream from the server to the client
+fn add_benchmark_group(benchmarks: &mut Vec<Benchmark>, param: BenchmarkParam) {
+    let tag = &param.tag.clone();
+
+    // Handshake benchmarks
+    let all_resumption_params = [
+        ResumptionParam::No,
+        ResumptionParam::SessionID,
+        ResumptionParam::Tickets,
+    ];
+
+    let param = Arc::new(param);
+
+    for resumption_param in all_resumption_params {
+        let captured_param1 = param.clone();
+        let captured_param2 = param.clone();
+        benchmarks.push(
+            Benchmark::new(
+                format!("new_conn_{}_{tag}", resumption_param.label()),
+                move |_reader, _writer| {
```

Comment:
The `SideFn` stuff seems pretty hard to understand to me. Could we unroll this into an actual trait with two implementing types?

Reply:
I'm not sure how a trait would fit here, could you expand on the aspects that are currently hard to understand?

For context, each side of a benchmark (client and server) runs in its own subprocess. To make this possible, each benchmark must specify how to run each side. A `SideFn` is a function that runs one side of the benchmark: it needs access to a reader and a writer, and it returns a result.

Reply:
In practice, the `SideFn` abstraction seems to be abstracting over (a) running `make_client_config()` vs `make_server_config()` and (b) `handshake_client_side()` vs `handshake_server_side()`. If that's the goal of the abstraction, I think it would be clearer to have a trait that has these two required methods. There's also a bunch of duplication between the two `SideFn` implementations which would be nice to reduce if possible.

Reply:
I see. I got a bit carried away by the desire to keep the benchmarking stuff generic, but now I realize it makes the code more difficult to read. I have replaced the usage of closures with an enum to specify which code should be run for the benchmark (i.e. only the handshake vs. handshake + transfer). Haven't pushed yet but will give you a head's up when I do.